src/import/sst_importer.rs

```diff
@@ -0,0 +1,402 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::fs::{self, File, OpenOptions};
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use crc::crc32::{self, Hasher32};
+use uuid::Uuid;
+use kvproto::importpb::*;
+
+use util::collections::HashMap;
+
+use super::{Error, Result};
+
+pub type Token = usize;
+
+pub struct SSTImporter {
+    dir: ImportDir,
+    token: AtomicUsize,
+    files: Mutex<HashMap<Token, ImportFile>>,
+}
+
+impl SSTImporter {
+    pub fn new<P: AsRef<Path>>(root: P) -> Result<SSTImporter> {
+        Ok(SSTImporter {
+            dir: ImportDir::new(root)?,
+            token: AtomicUsize::new(1),
+            files: Mutex::new(HashMap::default()),
+        })
+    }
+
+    pub fn token(&self) -> Token {
+        self.token.fetch_add(1, Ordering::SeqCst)
+    }
+
+    fn insert(&self, token: Token, file: ImportFile) {
+        let mut files = self.files.lock().unwrap();
+        assert!(files.insert(token, file).is_none());
+    }
+
+    pub fn remove(&self, token: Token) -> Option<ImportFile> {
+        let mut files = self.files.lock().unwrap();
+        files.remove(&token)
+    }
+
+    pub fn create(&self, token: Token, meta: &SSTMeta) -> Result<()> {
+        let mut files = self.files.lock().unwrap();
+        if files.contains_key(&token) {
+            return Err(Error::TokenExists(token));
+        }
+
+        match self.dir.create(meta) {
+            Ok(f) => {
+                info!("create {:?}", f);
+                files.insert(token, f);
+                Ok(())
+            }
+            Err(e) => {
+                error!("create {:?}: {:?}", meta, e);
+                Err(e)
+            }
+        }
+    }
+
+    pub fn append(&self, token: Token, data: &[u8]) -> Result<()> {
+        match self.remove(token) {
+            Some(mut f) => match f.append(data) {
+                Ok(_) => {
+                    self.insert(token, f);
```

Comment:
Since it needs to be inserted back, how about `HashMap::entry`? It only hashs once.

Reply:
It can't because we need to append data without the lock.